util/topsql/reporter/reporter.go

```diff
@@ -15,626 +15,274 @@
-		entry, exist := collectTarget[key]
-		if !exist {
-			entry = &dataPoints{
-				SQLDigest:     record.SQLDigest,
-				PlanDigest:    record.PlanDigest,
-				CPUTimeMsList: make([]uint32, 1, listCapacity),
-				TimestampList: make([]uint64, 1, listCapacity),
-			}
-			entry.CPUTimeMsList[0] = record.CPUTimeMs
-			entry.TimestampList[0] = timestamp
-			collectTarget[key] = entry
-		} else {
-			entry.CPUTimeMsList = append(entry.CPUTimeMsList, record.CPUTimeMs)
-			entry.TimestampList = append(entry.TimestampList, timestamp)
-		}
-		entry.CPUTimeMsTotal += uint64(record.CPUTimeMs)
-	}
-
 	if len(evicted) == 0 {
 		return
 	}
-	// Merge non Top N data as "others" (keyed by in the `keyOthers`) in `collectTarget`.
-	// SQL meta will not be evicted, since the evicted SQL can be appear on Other components (TiKV) TopN records.
 	totalEvictedCPUTime := uint32(0)
-	for _, evict := range evicted {
-		totalEvictedCPUTime += evict.CPUTimeMs
+	for _, e := range evicted {
+		totalEvictedCPUTime += e.CPUTimeMs
+		// Mark which digests are evicted under each timestamp.
+		// We will determine whether the corresponding CPUTime has been evicted
+		// when collecting stmtstats. If so, then we can ignore it directly.
+		tsr.collecting.markAsEvicted(timestamp, e.SQLDigest, e.PlanDigest)
+	}
+	tsr.collecting.appendOthersCPUTime(timestamp, totalEvictedCPUTime)
+}
+
+// doCollectStmtStatsMaps collects tsr.stmtStatsMaps into tsr.collecting.
+// All the evicted items will be summary into the others.
+func (tsr *RemoteTopSQLReporter) doCollectStmtStatsMaps() {
+	defer util.Recover("top-sql", "doCollectStmtRecords", nil, false)
+
+	for timestamp, data := range tsr.stmtStatsMaps {
+		for digest, item := range data {
+			sqlDigest, planDigest := []byte(digest.SQLDigest), []byte(digest.PlanDigest)
+			if tsr.collecting.hasEvicted(timestamp, sqlDigest, planDigest) {
+				// This timestamp+sql+plan has been evicted due to low CPUTime.
+				tsr.collecting.appendOthersStmtStatsItem(timestamp, *item)
+				continue
+			}
+			tsr.collecting.getOrCreateRecord(sqlDigest, planDigest).appendStmtStatsItem(timestamp, *item)
+		}
 	}
-	addEvictedCPUTime(collectTarget, timestamp, totalEvictedCPUTime)
+	tsr.stmtStatsMaps = map[uint64]stmtstats.StatementStatsMap{}
 }
 
-// takeDataAndSendToReportChan takes collected data and then send to the report channel for reporting.
-func (tsr *RemoteTopSQLReporter) takeDataAndSendToReportChan(collectedDataPtr *map[string]*dataPoints) {
-	data := collectedData{
-		records:           *collectedDataPtr,
-		normalizedSQLMap:  tsr.normalizedSQLMap.Load().(*sync.Map),
-		normalizedPlanMap: tsr.normalizedPlanMap.Load().(*sync.Map),
+func (tsr *RemoteTopSQLReporter) takeDataFromCollectChanBuffer() {
+	timestamp := uint64(nowFunc().Unix())
+	for {
+		select {
+		case data := <-tsr.collectCPUTimeChan:
+			tsr.doCollect(timestamp, data)
+		case data := <-tsr.collectStmtStatsChan:
+			tsr.stmtStatsMaps[timestamp] = data
+		default:
+			return
+		}
```

Comment:
Will `select` respect the cases order?

Reply:
No, we don’t depend on this. Data from `collectCPUTimeChan` will be directly processed by `tsr.doCollect()`, and data from `collectStmtStatsChan` will be processed **lazily** by `tsr.doCollectStmtStatsMaps()` when reportTicker expires. Therefore, the StmtStats data must be processed after the CPU data.

Reply:
So we should introduce two loops, one for collecting cpu records, one for collecting stmt records.

Reply:
Oh, got you.